### ISSUE TITLE
systemd service stop doesn't drop memory utilization to 0

### ISSUE THREAD

**Issue by tycho** [firehol/netdata#1586]
I was watching the services.mem_usage graph and stopped a service. The memory usage stayed constant at a non-zero value until I started it again.

I think what happens here is that we don't record a zero value when we first notice a service go missing.

**Comment by tycho** [comment]
This is kind of a hack, but it fixes the problem:
```diff
diff --git a/src/common.c b/src/common.c
index 36fa0c9..79d13ba 100644
--- a/src/common.c
+++ b/src/common.c
@@ -1172,18 +1172,30 @@ void get_system_HZ(void) {
 }
 
 int read_single_number_file(const char *filename, unsigned long long *result) {
+    int err;
+
     char buffer[1024 + 1];
 
     int fd = open(filename, O_RDONLY, 0666);
-    if(unlikely(fd == -1)) return 1;
+    if(unlikely(fd == -1)) {
+        err = 1;
+        goto fail;
+    }
 
     ssize_t r = read(fd, buffer, 1024);
     if(unlikely(r == -1)) {
         close(fd);
-        return 2;
+        err = 2;
+        goto fail;
     }
 
     close(fd);
     *result = strtoull(buffer, NULL, 0);
     return 0;
+fail:
+    if (*result != 0) {
+        *result = 0;
+        return 0;
+    }
+    return err;
 }
```
Basically, the first time we error, we set the result to zero and call it a success. Every subsequent failure just returns a nonzero error code.

 Maybe we need a special wrapper for `read_single_number_file` which is something like `read_single_number_file_or_zero` which does as the above version does in the `fail` block.

**Comment by ktsaou** [comment]
I think this is a visualization problem. The value is null (`NaN`) on the chart, but when you hover on the chart the legend is not refreshed to show the fact (remembers the last value rendered on the chart).

Can you verify it?

**Comment by tycho** [comment]
Proposed wrapper to replace existing `read_single_number_file` callsites in `sys_fs_cgroup.c`:

```diff
diff --git a/src/inlined.h b/src/inlined.h
index 87a54dd..2156034 100644
--- a/src/inlined.h
+++ b/src/inlined.h
@@ -72,4 +72,21 @@ static inline int read_single_number_file(const char *filename, unsigned long lo
     return 0;
 }
 
+static inline int read_single_number_file_or_zero(const char *filename, unsigned long long *result) {
+       int ret = read_single_number_file(filename, result);
+
+       /* Success */
+       if (!ret)
+               return ret;
+
+       /* First failure, write zero into *result */
+       if (*result != 0) {
+               *result = 0;
+               return 0;
+       }
+
+       /* Subsequent failures, return the error code. */
+       return ret;
+}
+
 #endif //NETDATA_INLINED_H
```

**Comment by tycho** [comment]
So I know what I saw wasn't just the visualization problem you describe because it showed up in both the chart and in the legend.

**Comment by tycho** [comment]
Look here:

https://netdata.uplinklabs.net/#menu_services_submenu_mem;after=1484694805000;before=1484696890000;theme=slate;help=false

Filter to only show nagios, and look between timestamps 15:32:50 to 15:35:00 -- the service was stopped during that time, and the usage was nonzero in both the chart and in the legend. I verified the service wasn't running and that the cgroup was gone.

**Comment by tycho** [comment]
I am pretty sure the reason is that callers of read_single_number_file only report that the value is "updated" when the function call succeeds (hence the wrapper I provided above)

**Comment by ktsaou** [comment]
for reference: fixed with #1584
